src/crimson/net/Socket.h

```diff
@@ -0,0 +1,86 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2018 Intel Corp.
+ *
+ * Author: Yingxin Cheng <yingxincheng@gmail.com>
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#pragma once
+
+#include <memory>
+
+#include <seastar/core/reactor.hh>
+#include <seastar/net/packet.hh>
+
+#include "include/buffer.h"
+#include "Errors.h"
+
+namespace ceph::net {
+
+class Socket
+{
+  seastar::connected_socket socket;
+  seastar::input_stream<char> in;
+  seastar::output_stream<char> out;
+
+  bool is_closed = false;
+  bool allow_destruct = false;
+
+  /// buffer state for read()
+  struct Reader {
+    bufferlist buffer;
+    size_t remaining;
+  } r;
+
+ public:
+  explicit Socket(seastar::connected_socket&& _socket)
+    : socket(std::forward<seastar::connected_socket>(_socket)),
+      in(socket.input()),
+      out(socket.output()) {}
+  Socket(Socket&& o)
+    : socket(std::move(o.socket)),
+      in(std::move(o.in)),
+      out(std::move(o.out)),
+      r(std::move(o.r)) {
+    ceph_assert(!o.is_closed);
```

Comment:
Because we don't replace a closed connection in this implementation, thus a closed `Socket` is not assumed to be moved, if it happens, it could be a bug.